### PR TITLE
Add Fireworks to LlamaIndex

### PR DESCRIPTION
**Pull request by benjibc**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change
- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- Added new unit/integration tests

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/base.py
Comment (logan-markewich) [inline llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/base.py:1]: I might be missing something, but I think this is very nearly a copy of the openai embedding class.

Would it be possible to add `llama-index-embeddings-openai` to the dependencies and then extend that class? I think a large majority of code can be reused? 

Let me know if that's not the case

**Comment by benjibc** [comment]
@logan-markewich updated to use OpenAI embedding, but I had to refactor OpenAI embedding's base class to handle different credentials, so please take a look and see if the change make sense to you.

File: llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/base.py
Comment (logan-markewich) [reply llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/base.py:1]: Perfect!

File: docs/examples/embeddings/fireworks.ipynb
Comment (logan-markewich) [inline docs/examples/embeddings/fireworks.ipynb:1]: Ok last request! Can you also add a notebook for the LLM class as well?

And then add links to the embeddings in `docs/module_guides/models/embeddings.md` and LLM linked in `docs/module_guides/models/llms/modules.md`

File: llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/utils.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+from typing import Optional, Tuple
+
+from llama_index.core.llms.generic_utils import get_from_param_or_env
```
Comment (logan-markewich) [inline llama-index-integrations/embeddings/llama-index-embeddings-fireworks/llama_index/embeddings/fireworks/utils.py:3]: Should be `from llama_index.core.base.llms.generic utils import get_from_param_or_env`

File: llama-index-integrations/llms/llama-index-llms-fireworks/llama_index/llms/fireworks/base.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+from typing import Any, Callable, Dict, Optional, Sequence
+
+from llama_index.core.base.llms.types import ChatMessage, LLMMetadata
+from llama_index.core.callbacks import CallbackManager
+from llama_index.core.constants import DEFAULT_NUM_OUTPUTS, DEFAULT_TEMPERATURE
+from llama_index.core.llms.generic_utils import get_from_param_or_env
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-fireworks/llama_index/llms/fireworks/base.py:6]: Should be `from llama_index.core.base.llms.generic_utils import get_from_param_or_env`

File: llama-index-integrations/llms/llama-index-llms-fireworks/llama_index/llms/fireworks/utils.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
+from typing import Any, Dict, List, Optional, Sequence, Tuple
+
+from llama_index.core.base.llms.types import ChatMessage, MessageRole
+from llama_index.core.llms.generic_utils import get_from_param_or_env
```
Comment (logan-markewich) [inline llama-index-integrations/llms/llama-index-llms-fireworks/llama_index/llms/fireworks/utils.py:4]: Should be `from llama_index.core.base.llms.generic_utils import get_from_param_or_env`

**Review by logan-markewich — APPROVED** [review]

**Comment by benjibc** [comment]
^ fixed unittest with the wrong import

**Comment by ckrapu-nv** [comment]
This PR added `fireworks` as a dependency for multiple workflows that shouldn't even touch them. For example, anything using `LangChainLLM` now requires `llama-index-llms-fireworks`. This is leading to dependency management issues for our projects since we can't resolve a satisfactory version of `llama-index-llms-fireworks` with our other pinned dependencies. My recommendation is to add the imports within the conditional logic, not at the module level.

See [here](https://github.com/run-llama/llama_index/blame/3c003c957144d011b3b60359572204a3eb1567dc/llama-index-integrations/llms/llama-index-llms-langchain/llama_index/llms/langchain/utils.py#L6).

@benjibc

**Comment by logan-markewich** [comment]
@ckrapu-nv fixed when this merges https://github.com/run-llama/llama_index/pull/14893

**Pull request merged**